blueocean-web/src/main/java/io/jenkins/blueocean/ResourceManager.java

```diff
@@ -0,0 +1,135 @@
+package io.jenkins.blueocean;
+
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import hudson.Plugin;
+import hudson.PluginManager;
+import hudson.PluginWrapper;
+import hudson.Util;
+import jenkins.model.Jenkins;
+import jnr.ffi.annotations.In;
+import org.apache.commons.lang.StringUtils;
+import org.kohsuke.stapler.HttpResponses;
+import org.kohsuke.stapler.MetaClass;
+import org.kohsuke.stapler.StaplerRequest;
+import org.kohsuke.stapler.StaplerResponse;
+
+import javax.servlet.ServletContext;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import java.io.IOException;
+import java.net.URL;
+import java.util.Arrays;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+
+import static javax.servlet.http.HttpServletResponse.SC_BAD_REQUEST;
+import static javax.servlet.http.HttpServletResponse.SC_FORBIDDEN;
+import static javax.servlet.http.HttpServletResponse.SC_NOT_FOUND;
+
+/**
+ * Plugin-aware resource management system
+ */
+@Singleton
+public class ResourceManager {
+
+    /**
+     * Map used as a set to remember which resources can be served.
+     */
+    private final ConcurrentHashMap<String,String> allowedResources = new ConcurrentHashMap<>();
+
+    @Inject
+    private PluginManager pluginManager;
+
+    @Inject
+    private ServletContext context;
+
+    private static final String KEY = ResourceManager.class.getName();
+
+    public ResourceManager() {
+        install();
+    }
+
+    /**
+     * Handle URLs like /resources/my-plugin/c82be88/org/jenkins-ci/plugins/bundle.js
```

Comment:
@tfennelly the URL format is as follows

`/resources/my-plugin/c82be88/org/jenkins-ci/plugins/bundle.js`

# `/resources/` gets served off of our root action
# `my-plugin` is the name of the plugin the resource belongs to
# `c82be88` is a hash of the plugin version and the jenkins version. That means if either of those change new resources are downloaded.
# `org/jenkins-ci/plugins/bundle.js` is the resource path within the plugin JAR